distributed/shuffle/shuffle_extension.py

```diff
@@ -27,93 +25,62 @@
-
-    def _partition_range(self, worker: str) -> tuple[int, int]:
-        "Get the output partition numbers (inclusive) that a worker will hold"
-        i = self.workers.index(worker)
-        first = math.ceil(self.npartitions * i / len(self.workers))
-        last = math.ceil(self.npartitions * (i + 1) / len(self.workers)) - 1
-        return first, last
-
-    def npartitions_for(self, worker: str) -> int:
-        "Get the number of output partitions a worker will hold"
-        first, last = self._partition_range(worker)
-        return last - first + 1
-
-
 class Shuffle:
-    "State for a single active shuffle"
+    """State for a single active shuffle"""
 
     def __init__(
         self,
-        metadata: ShuffleMetadata,
+        column,
+        workers: list[str],
+        worker_for: dict[int, str],
+        schema: pa.Schema,
+        id: ShuffleId,
         worker: Worker,
         executor: ThreadPoolExecutor,
     ) -> None:
-        self.metadata = metadata
+
+        import pandas as pd
+        import pyarrow as pa
+
+        self.column = column
+        self.id = id
+        self.schema = schema
         self.worker = worker
+        self.all_workers = workers
         self.executor = executor
 
-        import pyarrow as pa
+        partitions_of = defaultdict(list)
+        for part, address in worker_for.items():
+            partitions_of[address].append(part)
+        self.partitions_of = dict(partitions_of)
+        self.worker_for = pd.Series(worker_for, name="_workers").astype("category")
 
         self.multi_file = MultiFile(
             dump=functools.partial(
-                dump_batch, schema=pa.Schema.from_pandas(self.metadata.empty)
+                dump_batch,
+                schema=self.schema,
             ),
             load=load_arrow,
-            directory=os.path.join(self.worker.local_directory, str(self.metadata.id)),
+            directory=os.path.join(self.worker.local_directory, str(self.id)),
             join=pa.concat_tables,  # pd.concat
             sizeof=lambda L: sum(map(len, L)),
-            loop=worker.io_loop,
+            loop=self.worker.io_loop,
         )
 
         async def send(address, shards):
             return await self.worker.rpc(address).shuffle_receive(
                 data=to_serialize(shards),
-                shuffle_id=self.metadata.id,
+                shuffle_id=self.id,
             )
 
         self.multi_comm = MultiComm(
             send=send,
-            loop=worker.io_loop,
+            loop=self.worker.io_loop,
         )
-        MultiComm.max_connections = min(len(self.metadata.workers), 10)
+        MultiComm.max_connections = min(len(self.partitions_of), 10)
 
         self.diagnostics: dict[str, float] = defaultdict(float)
-        self.output_partitions_left = metadata.npartitions_for(worker.address)
+        self.output_partitions_left = len(self.partitions_of[self.worker.address])
```

Comment:
TODO: we now the exact partitions now, maybe we want to maintain a set rather than a counter.